java/memory/src/main/java/org/apache/arrow/memory/DictionaryEncodeHashMap.java

```diff
@@ -0,0 +1,298 @@
+        int oldValue = e.value;
+        e.value = value;
+        return oldValue;
+      }
+    }
+
+    addEntry(hash, key, value, i);
+    return NULL_VALUE;
+  }
+
+
+  @Override
+  public int remove(K key) {
+    Entry<K> e = removeEntryForKey(key);
+    return (e == null ? NULL_VALUE : e.value);
+  }
+
+  void createEntry(int hash, K key, int value, int bucketIndex) {
+    Entry<K> e = table[bucketIndex];
+    table[bucketIndex] = new Entry<>(hash, key, value, e);
+    size++;
+  }
+
+  private int putForNullKey(int value) {
+    for (Entry<K> e = table[0]; e != null; e = e.next) {
+      if (e.key == null) {
+        int oldValue = e.value;
+        e.value = value;
+        return oldValue;
+      }
+    }
+    addEntry(0, null, value, 0);
+    return NULL_VALUE;
+  }
+
+  void addEntry(int hash, K key, int value, int bucketIndex) {
+    if ((size >= threshold) && (null != table[bucketIndex])) {
+      resize(2 * table.length);
+      hash = (null != key) ? hash(key) : 0;
+      bucketIndex = indexFor(hash, table.length);
+    }
+
+    createEntry(hash, key, value, bucketIndex);
+  }
+
+  void resize(int newCapacity) {
+    Entry[] oldTable = table;
+    int oldCapacity = oldTable.length;
+    if (oldCapacity == MAXIMUM_CAPACITY) {
+      threshold = Integer.MAX_VALUE;
+      return;
+    }
+
+    Entry[] newTable = new Entry[newCapacity];
+    transfer(newTable);
+    table = newTable;
+    threshold = (int)Math.min(newCapacity * loadFactor, MAXIMUM_CAPACITY + 1);
+  }
+
+  void transfer(Entry[] newTable) {
+    int newCapacity = newTable.length;
+    for (Entry<K> e : table) {
+      while (null != e) {
+        Entry<K> next = e.next;
+        int i = indexFor(e.hash, newCapacity);
+        e.next = newTable[i];
+        newTable[i] = e;
+        e = next;
+      }
+    }
+  }
+
+  final Entry<K> removeEntryForKey(Object key) {
+    if (size == 0) {
+      return null;
+    }
+    int hash = (key == null) ? 0 : hash(key);
+    int i = indexFor(hash, table.length);
+    Entry<K> prev = table[i];
+    Entry<K> e = prev;
+
+    while (e != null) {
+      Entry<K> next = e.next;
+      Object k;
+      if (e.hash == hash && ((k = e.key) == key || (key != null && key.equals(k)))) {
+        size--;
+        if (prev == e) {
+          table[i] = next;
+        } else {
+          prev.next = next;
+        }
+
+        return e;
+      }
+      prev = e;
+      e = next;
+    }
+
+    return e;
+  }
+
+  /**
+   * Returns the number of mappings in this Map.
+   */
+  public int size() {
+    return size;
+  }
+
+  /**
+   * Removes all elements from this map, leaving it empty.
+   */
+  public void clear() {
+    size = 0;
+    for (int i = 0; i < table.length; i++) {
+      table[i] = null;
+    }
+  }
+
+  /**
+   * Class to keep key-value data within hash map.
+   * @param <K> key type.
+   */
+  static class Entry<K> {
+    final K key;
+    int value;
+    Entry<K> next;
```

Comment:
open addressing can perform better under some circumstances I think but this is a good start.

Reply:
Agreed, I have tested ObjectIntMap in eclipse-collections which use open addressing but seems performance is worse than HashMap, BWT, some attempts can be made as follows. Thanks!